### PR TITLE
JWT expiration check

### PR DESCRIPTION
**Pull request by Scurrra**
### Motivation:

Doing a project I found this library for easy oauth. Because of lack of experience in the field I decided to learn how the library works. I had a big question on how jwts work, especially in this library. While doing this I found out that there is no check if the token is expired. I added this check to OAuth2Backend.authenticate. I also fixed deprecated datetime.utcnow() function.

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing](./CONTRIBUTING.md) document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change?

### Changes to Core Features:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you updated the documentation related to the changes you have made?
* [ ] Have you written new tests for your core changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?

### CODE REVIEW

File: src/fastapi_oauth2/middleware.py [hunk 1]
```diff
@@ -1,5 +1,6 @@
 from datetime import datetime
 from datetime import timedelta
+from datetime import UTC
```
Comment (ArtyomVancyan) [inline src/fastapi_oauth2/middleware.py:3]:
```suggestion
from datetime import timezone
```

File: src/fastapi_oauth2/middleware.py [hunk 2]
```diff
@@ -51,7 +53,7 @@ def jwt_decode(cls, token: str) -> dict:
 
     @classmethod
     def jwt_create(cls, token_data: dict) -> str:
-        expire = datetime.utcnow() + timedelta(seconds=cls.expires)
+        expire = datetime.now(UTC) + timedelta(seconds=cls.expires)
```
Comment (ArtyomVancyan) [inline src/fastapi_oauth2/middleware.py:56]:
```suggestion
        expire = datetime.now(timezone.utc) + timedelta(seconds=cls.expires)
```

File: src/fastapi_oauth2/middleware.py [hunk 3]
```diff
@@ -106,7 +108,11 @@ async def authenticate(self, request: Request) -> Optional[Tuple[Auth, User]]:
         if not scheme or not param:
             return Auth(), User()
 
-        user = User(Auth.jwt_decode(param))
+        token_data = Auth.jwt_decode(param)
+        if token_data["exp"] and token_data["exp"] < int(datetime.now(UTC).timestamp()):
```
Comment (ArtyomVancyan) [inline src/fastapi_oauth2/middleware.py:112]:
```suggestion
        if token_data["exp"] and token_data["exp"] < int(datetime.now(timezone.utc).timestamp()):
```

**Review by ArtyomVancyan — CHANGES_REQUESTED** [review]
Hi @Scurrra, thanks for pointing it out. I suggest changing the `UTC` alias with `timezone.utc` as it is not available in all versions of the `datetime` package.

**Comment by ArtyomVancyan** [comment]
@Scurrra, you requested me a review, but I requested a few changes before merging (you shouldn't click "Resolve", you should do "Commit suggestion"). I will wait for your update. Also, I didn't get what was wrong with the old `datetime.utcnow()`. Isn't that the same as what you suggested (`datetime.now(UTC)`)?

**Comment by Scurrra** [comment]
> you should do "Commit suggestion"
My bad, sorry.

datetime.utcnow() will be deprecated.

**Comment by Scurrra** [comment]
I don't know is it my fault or not, but playground is broken.

image

Probably this is because of troubles with OAUTH2_GITHUB_CLIENT_ID in playground, which wasn't affected by me.

**Comment by ArtyomVancyan** [comment]
It was a missdeployment issue, I have fixed it. Thanks for letting me know. Also, it could not be your fault, as these changes haven't been released yet :)

**Pull request merged**